webapp/src/main/java/com/box/l10n/mojito/service/image/S3ImageService.java

```diff
@@ -0,0 +1,69 @@
+package com.box.l10n.mojito.service.image;
+
+import com.box.l10n.mojito.entity.Image;
+import com.box.l10n.mojito.service.blobstorage.s3.S3BlobStorage;
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Service to upload and retrieve images from S3.
+ * <p>
+ * If a requested image cannot be found in S3 an attempt will be made to retrieve from the Mojito DB, if the image is
+ * found in the DB then an async task will be executed to upload the image to S3.
+ * <p>
+ * Configured {@link S3BlobStorage} and {@link com.amazonaws.services.s3.AmazonS3} client instances are required to
+ * upload and retrieve images from S3.
+ *
+ * @author maallen
+ */
+@Component
+@ConditionalOnProperty(value = "l10n.blob-storage.s3.images.enabled", havingValue = "true")
+public class S3ImageService implements ImageService {
+
+    /**
+     * logger
+     */
+    static Logger logger = getLogger(S3ImageService.class);
+
+    @Autowired
+    S3BlobStorage s3BlobStorage;
+
+    @Autowired
+    S3UploadImageAsyncTask s3UploadImageAsyncTask;
+
+    @Autowired
+    ImageRepository imageRepository;
+
+    @Override
+    public Image getImage(String name) {
```

Comment:
what about having have another service implementation: `S3FallbackImageService`. This way the `S3ImageService` stays clean.

S3FallbackImageService#get would just be getting from S3 implementation and then from database implementation (probably need to return an Optional<Image> in the interface) and optionally trigger the async task.